src/bson.c

```diff
@@ -1201,6 +1289,195 @@ static bool php_phongo_bson_state_parse_type(zval* options, const char* name, ph
 	return retval;
 } /* }}} */
 
+static void field_path_map_element_set_info(php_phongo_field_path_map_element* element, php_phongo_bson_typemap_types type, zend_class_entry* ce TSRMLS_DC)
+{
+	element->node_type = type;
+	element->node_ce   = ce;
+}
+
+static void map_add_field_path_element(php_phongo_bson_typemap* map, php_phongo_field_path_map_element* element)
+{
+	/* Make sure we have allocated enough */
+	if (map->field_paths.allocated < map->field_paths.size + 1) {
+		map->field_paths.allocated += PHONGO_FIELD_PATH_EXPANSION;
+		map->field_paths.map = erealloc(map->field_paths.map, sizeof(php_phongo_field_path_map_element) * map->field_paths.allocated);
+	}
+
+	map->field_paths.map[map->field_paths.size] = element;
+	map->field_paths.size++;
+}
+
+static php_phongo_field_path_map_element* field_path_map_element_alloc(void)
+{
+	php_phongo_field_path_map_element* tmp = ecalloc(1, sizeof(php_phongo_field_path_map_element));
+
+	tmp->entry = php_phongo_field_path_alloc(true);
+
+	return tmp;
+}
+
+bool php_phongo_bson_state_add_field_path(php_phongo_bson_typemap* map, char* field_path_original, php_phongo_bson_typemap_types type, zend_class_entry* ce TSRMLS_DC)
+{
+	char*                              saveptr;
+	char*                              field_path_str = estrdup(field_path_original);
+	const char*                        element;
+	php_phongo_field_path_map_element* field_path_map_element;
+
+	field_path_map_element = field_path_map_element_alloc();
+
+	/* Loop over all the segments. A segment is delimited by a "." */
+	element = php_strtok_r(field_path_str, ".", &saveptr);
+	while (element) {
```

Comment:
Is there a case where we don't enter the `while` loop on the first iteration?

It may be good to have error tests for field paths such as `"..."`, `".foo"`, and `""` so we can test how that is handled.

Reply:
I don't think we won't ever enter, but I will add test cases. I know that ``strtok`` handles ``...`` as ``.`` for example.

Reply:
Sigh, strtok_r is pretty useless, as it can't handle empty elements (like in ``foo..bar``), so I've had to write this, which also means that we needed to keep some ``TSRMLS_[DC]C`` things around.